jupyterhub/utils.py

```diff
@@ -296,6 +299,30 @@ def metrics_authentication(self):
         raise web.HTTPError(403)
 
 
+def needs_scope(scope):
+    """Decorator to restrict access to users or services with the required scope"""
+
+    def scope_decorator(func):
+        @functools.wraps(func)
+        def _auth_func(self, *args, **kwargs):
+            allows_subset = 'subset' in func.__code__.co_varnames
+            if scope in self.scopes:
+                return func(self, *args, **kwargs)
+            elif allows_subset:
+                # Check if access is not restricted to user/server/service
+                match_string = re.compile("^" + re.escape(scope) + r"!.+=.+$")
+                subscopes = filter(lambda s: re.search(match_string, s), self.scopes)
+                subset = [subscope.split('=')[1] for subscope in subscopes]
```

Comment:
maybe parse these to a dict for example `!user=name` would give `{"user": "name"}`

Reply:
We also need to be careful because the filter applies on a per-scope basis.


Let's add lots of unittests for just this decorator. We will have lots of combinations of multiple filters, etc.

Maybe add a utility for checking later on, maybe parse scopes to objects with methods (or utility functions) something like:


```python
check_scope(have_scopes, needed_scope, user=x, server=y, group=z) -> bool
```

Reply:
Maybe another utility for extract limited scope

```
extract_filtered_scope(scopes, needed_scope) -> subset dict
```